projects/robots/festo/robotino3/controllers/robotino3/robotino3.c

```diff
@@ -0,0 +1,274 @@
+  double sensors_values[NUMBER_OF_INFRARED_SENSORS];
+  for (int i = 0; i < NUMBER_OF_INFRARED_SENSORS; ++i)
+    sensors_values[i] = convert_volt_to_meter(infrared_sensors[0], wb_distance_sensor_get_value(infrared_sensors[i]));
+
+  if (display_message) {
+    // Display some IR sensor data every second
+    int display_second = (int)wb_robot_get_time();
+    if (display_second != *last_displayed_second) {
+      *last_displayed_second = display_second;
+      printf("time = %d [s]\n", display_second);
+      for (int i = 0; i < NUMBER_OF_INFRARED_SENSORS; ++i)
+        printf("infrared sensor('%s') = %f [m]\n", infrared_sensors_names[i], sensors_values[i]);
+    }
+  }
+  base_braitenberg_avoidance(sensors_values);
+}
+
+static void display_instructions() {
+  printf("Control commands:\n");
+  printf(" - Arrows:       Move the robot\n");
+  printf(" - Page Up/Down: Rotate the robot\n");
+  printf(" - Space: Stop/Reset\n");
+  printf("\n");
+  printf("Demonstration mode: press 'D'\n");
+  printf("Autopilot mode : press 'A'\n");
+}
+
+static void check_keyboard() {
+  int key = wb_keyboard_get_key();
+  if ((key >= 0) && key != old_key) {
+    switch (key) {
+      case WB_KEYBOARD_UP:
+        printf("Go forwards\n");
+        base_forwards();
+        autopilot = false;
+        break;
+      case WB_KEYBOARD_DOWN:
+        printf("Go backwards\n");
+        base_backwards();
+        autopilot = false;
+        break;
+      case WB_KEYBOARD_LEFT:
+        printf("Strafe left\n");
+        base_strafe_left();
+        autopilot = false;
+        break;
+      case WB_KEYBOARD_RIGHT:
+        printf("Strafe right\n");
+        base_strafe_right();
+        autopilot = false;
+        break;
+      case WB_KEYBOARD_PAGEUP:
+        printf("Turn left\n");
+        base_turn_left();
+        autopilot = false;
+        break;
+      case WB_KEYBOARD_PAGEDOWN:
+        printf("Turn right\n");
+        base_turn_right();
+        autopilot = false;
+        break;
+      case WB_KEYBOARD_END:
+      case ' ':
+        printf("Reset\n");
+        base_reset();
+        break;
+      case 'D':
+        if (key != old_key)  // perform this action just once
+          demo = !demo;
+        break;
+      case 'A':
+        if (key != old_key)  // perform this action just once
+          autopilot = !autopilot;
+        break;
+      default:
+        fprintf(stderr, "Wrong keyboard input\n");
+        break;
+    }
+  }
+  if (autopilot != old_autopilot) {
+    old_autopilot = autopilot;
+    if (autopilot)
+      printf("Auto control\n");
+    else
+      printf("Manual control\n");
+  }
+  old_key = key;
+}
+
+int main(int argc, char **argv) {
+  // Initialization
+  wb_robot_init();
+
+  // Get and enable the infrared sensors
+  for (int i = 0; i < NUMBER_OF_INFRARED_SENSORS; ++i) {
+    infrared_sensors[i] = wb_robot_get_device(infrared_sensors_names[i]);
+    wb_distance_sensor_enable(infrared_sensors[i], TIME_STEP);
+  }
+
```

Comment:
It would be nice to enable the camera too, this will make the demo more interesting.